compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -298,39 +314,45 @@ macro_rules! should_ever_cache_on_disk {
 
 pub(crate) fn create_query_frame<
     'tcx,
-    K: Copy + Key + for<'a> HashStable<StableHashingContext<'a>>,
+    K: Copy + DynSend + DynSync + Key + for<'a> HashStable<StableHashingContext<'a>> + 'tcx,
 >(
     tcx: TyCtxt<'tcx>,
     do_describe: fn(TyCtxt<'tcx>, K) -> String,
     key: K,
     kind: DepKind,
     name: &'static str,
-) -> QueryStackFrame {
-    // If reduced queries are requested, we may be printing a query stack due
-    // to a panic. Avoid using `default_span` and `def_kind` in that case.
-    let reduce_queries = with_reduced_queries();
-
-    // Avoid calling queries while formatting the description
-    let description = ty::print::with_no_queries!(do_describe(tcx, key));
-    let description = if tcx.sess.verbose_internals() {
-        format!("{description} [{name:?}]")
-    } else {
-        description
-    };
-    let span = if kind == dep_graph::dep_kinds::def_span || reduce_queries {
-        // The `def_span` query is used to calculate `default_span`,
-        // so exit to avoid infinite recursion.
-        None
-    } else {
-        Some(key.default_span(tcx))
-    };
+) -> QueryStackFrame<QueryStackDeferred<'tcx>> {
     let def_id = key.key_as_def_id();
-    let def_kind = if kind == dep_graph::dep_kinds::def_kind || reduce_queries {
-        // Try to avoid infinite recursion.
-        None
-    } else {
-        def_id.and_then(|def_id| def_id.as_local()).map(|def_id| tcx.def_kind(def_id))
+
+    let extra = move || {
```

Comment:
if you give this closure a `TyCtxt` argument, the closure doesn't have to capture it or anything with such a lifetime anymore. And the only call site has access to a `TyCtxt`. This way you should be able to avoid any issues with the `dyn Fn` needing a `'tcx` bound

Reply:
Ah right. Since it is also `Copy`, can we avoid the dropck issues you encountered by using a `Copy` struct with a method instead of a closure?

Reply:
Not easily, as `QueryStackDeferred` is not generic over the key.

Reply:
Oof right. OK, then let's move the unsafety entirely into `QueryStackDeferred::new` by making it take a generic Copy parameter and having it allocate the arc and do the unsizing